### PR TITLE
arch-riscv: adding support for local interrupts

### PR DESCRIPTION
**Pull request by robhau**
Besides the standard RISC-V interrupts software, timer, and external interrupt, the RISC-V specification also offers the possibility to implement local interrupts. With this patch, we contribute an extension of RiscvInterrupts that enables connecting interrupt sources to the local interrupt controller. We assigned the local interrupts to machine-level and gave them the highest priority. If two local interrupts are pending, there exception code will be the tie-breaker (higher ID > lower ID). 32 Bit systems only recognize the local interrupts 16 to 31, 64 Bit systems 16 to 63.

Change-Id: Iff8d34e740b925dce351c0c6f54f4bd37a647e0c

### CODE REVIEW

File: src/arch/riscv/RiscvInterrupts.py [hunk 1]
```diff
@@ -28,9 +29,34 @@
...
     cxx_class = "gem5::RiscvISA::Interrupts"
     cxx_header = "arch/riscv/interrupts.hh"
+
+    local_interrupt_pins = VectorIntSinkPin("Pins for local interrupts")
+    local_interrupt_ids = VectorParam.Unsigned(
+        [], "list of local interrupt ids"
+    )
+
+
+add_citation(
```
Comment (powerjg) [inline src/arch/riscv/RiscvInterrupts.py:48]: Awesome! I'm glad someone is using this 😄

File: src/arch/riscv/interrupts.cc [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            INT_LOCAL_27, INT_LOCAL_26, INT_LOCAL_25, INT_LOCAL_24,
+            INT_LOCAL_23, INT_LOCAL_22, INT_LOCAL_21, INT_LOCAL_20,
+            INT_LOCAL_19, INT_LOCAL_18, INT_LOCAL_17, INT_LOCAL_16,
+            INT_EXT_MACHINE, INT_SOFTWARE_MACHINE, INT_TIMER_MACHINE,
+            INT_EXT_SUPER, INT_SOFTWARE_SUPER, INT_TIMER_SUPER,
+            INT_EXT_USER, INT_SOFTWARE_USER, INT_TIMER_USER
+        };
+        for (const int &id : interrupt_order)
+        if (checkInterrupt(id) && mask[id])
+            return std::make_shared<InterruptFault>(id);
```
Comment (powerjg) [inline src/arch/riscv/interrupts.cc:142]: Please add some curly braces here. It's quite unclear what the intention of this code is, and it's likely to cause future bugs.

File: src/arch/riscv/interrupts.cc [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
```
Comment (powerjg) [inline src/arch/riscv/interrupts.cc:155]: Curly braces here too, please

File: src/arch/riscv/interrupts.cc [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+Port &
+Interrupts::getPort(const std::string &if_name, PortID idx)
+{
+
+    if (if_name == "local_interrupt_pins" && idx < localInterruptPins.size()) {
+        return *localInterruptPins[idx];
+    } else {
+    return BaseInterrupts::getPort(if_name, idx);
```
Comment (powerjg) [inline src/arch/riscv/interrupts.cc:223]: Missing 4 leading spaces :)

**Review by powerjg — COMMENTED** [review]
Overall, this looks good to me. @rogerchang23424 could you take a look?

**Comment by robhau** [comment]
Thank you for your quick response. A colleague pointed out to me that i branched from stable, not from develop. I will fix that first. Then, I will add your suggestions.

**Comment by robhau** [comment]
I have some troubles with pushing my local changes into my forked gem5 repo. I will close this pullrequest and delete my remote branch. After pushing my local branch to github, i will reopen this pull request.

**Pull request closed without merging**